### ISSUE TITLE
i DID NOT UNDERSTAND WHERE THE PREFERENCES ARE INCLUDED ?

### ISSUE THREAD

**Issue by anwarghammam** [johan-kallstrom/gym-mo#1]
hello,
thank you for providing the code source of the project, One thing I did not understand, how are you including the preferences in the choice of the action ?

**Comment by johan-kallstrom** [comment]
In the work these envs (Gathering & Traffic) were used, this was done by learning (using modified DQN) a Q function that is conditioned on the agent's preferences: Q(s, p, a). So basically [s, p] is an augmented state space with the agent's preference vector in addition to the grid state. This agent was trained by uniformly sampling a new preference vector before each episode, to explore the preference space of the agent as well as the state space of the environment. We handle the preferences as part of the agent since different agents acting in the same environment can have different opinions about what is good or bad. In the Q function neural network the state observation was preprocessed (using image observation and CNN), and then concatenated with the preference vector for further processing to produce Q values.

**Comment by anwarghammam** [comment]
thank yo for  your quick answer, where the DQN code is implemented in this code source please ?

**Comment by johan-kallstrom** [comment]
The source for the agent is not in this repo, and it was not made public since we did not go ahead to formally publish the paper. However, you can find a similar, but more advanced approach in work done concurrently to ours in this repo: https://github.com/axelabels/DynMORL

**Comment by anwarghammam** [comment]
Thank you so much, 
As I understood from this repo is that the agent actions are related to the goal object index by just calculating the minimal distance between its position and the goal object. is that correct ? so there is no using of the preferences here right ?

**Comment by johan-kallstrom** [comment]
Correct, those calculations are for the agents that are considered part of the environment, which the learning agent is supposed to interact with. In the envrionment the preferences are only used to be able to set done=True in 'step' if there are no items/goals left that the learning agent is interested in.

**Comment by anwarghammam** [comment]
I got your point, for the usage of the preferences to set the done to True, I am still unable to understand their usage, can you help me with that ?

**Comment by johan-kallstrom** [comment]
The agents in 'gridworld_base.py' (HunterAgent, ScrollerAgent) are part of the environment and act according to some heuristic to move around in various ways, collect objects etc. So that code is unrelated to preferences, which are for external agents. All agent code and action selection is outside this repo.

The external (multi-objective) reinforcement learning agent is expected to act to maximise the utility of the expected return u(V). If you look in 'step' in MOGridworld in mo_gridworld_base.py, you can see that the reward is a vector (instead of the more common scalar reward) related to various events that occur in the environment (e.g., collect an item of a certain color, collide with a wall, collide with another agent). This means that the return V and Q function are also vectors, and to compare different vector valued returns, action values and rewards we use a linear utility function corresponding to the preferences to turn the vectors to scalars. So for instance reward_scalar = p1*r1 + p2*r2 + ... + pn*rn & p = [p1, ..., pn] for n objectives that the agent is interested in. So agents with different preferences p will select different actions, since they have different Q functions.

You can compare with the paper that uses the environments: http://www.diva-portal.org/smash/get/diva2:1362933/FULLTEXT01.pdf 

And our recent guide paper provides general information about multi-objective learning: https://link.springer.com/article/10.1007/s10458-022-09552-y

**Comment by johan-kallstrom** [comment]
In the above I should have said that the agents have different Q functions after scalarisation using the preferences.

**Comment by anwarghammam** [comment]
I got it thank you so much for your help !

**Comment by anwarghammam** [comment]
I have one last question, i don't understand the conditions to consider the episode as done (it is either we reached the maximum number of steps) or other conditions that i don't understand

**Comment by johan-kallstrom** [comment]
In the function is_done(...), there are two condtions that affect the implemented environments:

#1: When we reach a maximum number of allowed steps we are done
        if self.step_count > self.max_steps:
            return True

#2: If there are objects left to collect in the grid that will give a positive reward for the reinforcement learning agent in any objective according to its current preferences we are not done
        # if there is an object in this cell that would give a positive reward we are not done
        if self.grid[row, column] is not None and self.preference[self.grid[row, column].idx] > 0:
              return False

Then there are two additional conditions that do not affect the current environments, but could be used when constructing new environments:

#3: If there are objects left to collect in the grid that are desired by any of the heuristic agents of the environment we are not done. This could be used for cooperative scenarios when the reinforcement learning agent should learn to help the heuristic agents.
         # if there is an object in this cell  with and index that was specified as desired by heuristic agents (in the agent_preferences list given when constructing the env) we are not done
         if self.grid[row, column] is not None and self.grid[row, column].idx in agent_preferences:
               return False

#4: If there are heuristic agents left in the grid that when caught (e.g. in a hunter-prey scenario) would give the reinforcement learning agent a positive reward in any objective according to its current preferences we are not done:
         # if the reinforcement learning agent would step into the same cell as the heuristic agent with this index it would receive a positive reward
         if self.preference[agent.idx] > 0:
                return False


Default result:
If none of the conditions 2-4 are fulfilled we are done.

**Comment by anwarghammam** [comment]
Thank you for your detailed answer it was very helpful, one last question, I noticed that my agent is taken random initial position for every episode, I want that position to be fixed ? how can I do that ?

**Comment by johan-kallstrom** [comment]
The agent initial position should be fixed, maybe you are not calling env.reset() between episodes to reset the environment to the initial state?

**Comment by anwarghammam** [comment]
I developed more the moRandomPlayer and I want it now to be the default agent of my mo_gathering environment so I am doing the following : 
if __name__ == "__main__":
    my_grid = MOGridworld(
        gridworld_base.TEST_MAP,
        gridworld_base.TEST_MAPPING,
        preference=np.array([0.05, 0.05, 0.3, 0.3, 0.3]),
    )
    # 0.05,0.05,0.3,0.3,0.3
    # 0.05,0.05,0.7,0.1,0.1

    my_grid.reset()
    player = MOPlayer(my_grid)
    player.play()
I am not able to change the color of the agent neither keeping the same initial position when resetting the env.
it seems like this agent that i created is not the env agent ? how can I solve that ?

**Comment by johan-kallstrom** [comment]
The initial agent position is in the lower left corner, or specified when constructing the env, and is set on reset of the env (when episode results are printed to the terminal, and where there may be a small delay). 

Agent color for rendering was hard-coded to blue, but I updated it so that it can be set:

from gym_mo.envs.gridworlds.mo_gridworld_base import MORandomPlayer
from gym_mo.envs.gridworlds.mo_gathering_env import MOGatheringEnv

my_grid = MOGatheringEnv(agent_start=[0,0], agent_color=(255.0, 0.0, 255.0))
player = MORandomPlayer(my_grid, 10)
player.run()

**Comment by anwarghammam** [comment]
hello again, first thank you so much for your very detailed and quick answers. i have a question, I am trying a linear regression for the multiobjective agent, i need to give a presentation of the env that i will pass as features to the linear regression, I used the discreate observation of the env but it is not giving good results. any suggestions for that ?

**Comment by johan-kallstrom** [comment]
I have used the discrete observation with a neural network estimator, so it should be possible. Perhaps you have insufficient data/exploration? For the neural network approach some ~50000 episodes are required for convergence with a fixed preference over objectives. I am not sure what would be required for a linear regression approach.
